### PR TITLE
xDS router performance optimization

### LINKED ISSUE #1680: the xds router performance needs to be improved
### What would you like to be enhanced?

optimize the performance of xds router

### Why is this needed?

the xds router performance is optimized to reduce the extra CPU usage and invoking delay.

### PR DESCRIPTION
**Pull request by daizhenyu**
**What type of PR is this?**

Refactor.

**What this PR does / why we need it?**

The code logic of xds router is optimized to improve performance.
The extra CPU usage decreases to 1.5%, and the call delay decreases to 0.1 ms.

**Which issue(s) this PR fixes？**

Fixes #1680

**Does this PR introduce a user-facing change?**

No

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/sermant-io/Sermant/issues) related with this PR before you start working on it.
- [x] Make sure you have squashed your change to one single commit.
- [ ] GitHub Actions works fine in this PR.

### CODE REVIEW

File: sermant-plugins/sermant-router/router-common/src/main/java/io/sermant/router/common/utils/XdsRouterUtils.java [hunk 1]
```diff
@@ -88,6 +91,15 @@ public static void updateServiceDiscovery(XdsServiceDiscovery xdsServiceDiscover
         }
     }
 
+    /**
+     * updateServiceDiscovery
```
Comment (lilai23) [inline sermant-plugins/sermant-router/router-common/src/main/java/io/sermant/router/common/utils/XdsRouterUtils.java]: wrong comment

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/utils/BaseHttpRouterUtils.java [hunk 2]
```diff
@@ -110,35 +91,42 @@ public static Optional<ServiceInstance> chooseServiceInstanceByXds(String servic
...
-                .collect(Collectors.toMap(
-                        Entry::getKey,
-                        entry -> CollectionUtils.isEmpty(entry.getValue()) ? "" : entry.getValue().get(0)
-                ));
+        Map<String, String> processedHeaders = new HashMap<>();
+        for (Entry<String, List<String>> headerEntry : headers.entrySet()) {
+            processedHeaders.put(headerEntry.getKey(),
+                    CollectionUtils.isEmpty(headerEntry.getValue()) ? "" : headerEntry.getValue().get(0));
+        }
+        return processedHeaders;
```
Comment (lilai23) [inline sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/utils/BaseHttpRouterUtils.java]: headerEntry.getValue() is probably null

File: sermant-plugins/sermant-router/config/config.yaml [hunk 3]
```diff
@@ -3,6 +3,8 @@ router.plugin:
   enabled-registry-plugin-adaptation: false
   # whether to use xds route
   enabled-xds-route: false
+  # whether to use xds route for httpclient, httpasyncclient, okhttp, httpurlconnection
+  enabled-simple-http-route: false
```
Comment (hanbingleixue) [inline sermant-plugins/sermant-router/config/config.yaml]: enable-http-xds-route

File: sermant-plugins/sermant-router/router-common/src/main/java/io/sermant/router/common/utils/XdsRouterUtils.java (hunk 1 above)
Comment (daizhenyu) [reply sermant-plugins/sermant-router/router-common/src/main/java/io/sermant/router/common/utils/XdsRouterUtils.java]: done

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/utils/BaseHttpRouterUtils.java (hunk 2 above)
Comment (daizhenyu) [reply sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/utils/BaseHttpRouterUtils.java]: if headerEntry.getValue() is null or empty, put "" value into the map。Use CollectionUtils.isEmpty to check the value of header

Comment (daizhenyu) [reply sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/utils/BaseHttpRouterUtils.java]: done

**Review by lilai23 — DISMISSED** [review]

File: sermant-plugins/sermant-router/config/config.yaml (hunk 3 above)
Comment (daizhenyu) [reply sermant-plugins/sermant-router/config/config.yaml]: done

**Review by lilai23 — DISMISSED** [review]

**Review by lilai23 — DISMISSED** [review]

File: sermant-plugins/sermant-router/router-common/src/main/java/io/sermant/router/common/xds/XdsRouterHandler.java [hunk 4]
```diff
@@ -194,24 +183,47 @@ private Set<ServiceInstance> handleXdsRoute(XdsRoute route, String serviceName)
...
             XdsLocality locality) {
         return clusterLoadAssigment.getLocalityInstances().entrySet().stream()
-                .filter(xdsLocalitySetEntry -> xdsLocalitySetEntry.getKey().equals(locality))
+                .filter(xdsLocalitySetEntry -> isSameLocality(locality, xdsLocalitySetEntry.getKey()))
                 .flatMap(xdsLocalitySetEntry -> xdsLocalitySetEntry.getValue().stream())
                 .collect(Collectors.toSet());
     }
 
+    private boolean isSameLocality(XdsLocality selfLocality, XdsLocality serviceLocality) {
+        if (!selfLocality.getRegion().equals(serviceLocality.getRegion())) {
```
Comment (provenceee) [inline sermant-plugins/sermant-router/router-common/src/main/java/io/sermant/router/common/xds/XdsRouterHandler.java:192]: maybe selfLocality region is null?

File: sermant-plugins/sermant-router/router-common/src/main/java/io/sermant/router/common/xds/XdsRouterHandler.java [hunk 5]
```diff
@@ -194,24 +183,47 @@ private Set<ServiceInstance> handleXdsRoute(XdsRoute route, String serviceName)
...
+            return false;
+        }
+        if (StringUtils.isEmpty(selfLocality.getZone())) {
+            return true;
+        }
+        if (!selfLocality.getZone().equals(serviceLocality.getZone())) {
```
Comment (provenceee) [inline sermant-plugins/sermant-router/router-common/src/main/java/io/sermant/router/common/xds/XdsRouterHandler.java:198]: maybe selfLocality zone is null?

File: sermant-plugins/sermant-router/router-common/src/main/java/io/sermant/router/common/xds/XdsRouterHandler.java [hunk 6]
```diff
@@ -194,24 +183,47 @@ private Set<ServiceInstance> handleXdsRoute(XdsRoute route, String serviceName)
...
+            return false;
+        }
+        if (StringUtils.isEmpty(selfLocality.getSubZone())) {
+            return true;
+        }
+        return selfLocality.getSubZone().equals(serviceLocality.getSubZone());
```
Comment (provenceee) [inline sermant-plugins/sermant-router/router-common/src/main/java/io/sermant/router/common/xds/XdsRouterHandler.java:204]: maybe selfLocality subZone is null?

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/utils/BaseHttpRouterUtils.java [hunk 7]
```diff
@@ -110,35 +84,42 @@ public static Optional<ServiceInstance> chooseServiceInstanceByXds(String servic
...
+        if (StringUtils.isEmpty(serviceName) || serviceName.equals(LOCAL_HOST)
+                || !isStartWithLowercase(serviceName)) {
             return false;
         }
         return true;
     }
 
+    private static boolean isStartWithLowercase(String serviceName) {
+        char firstChar = serviceName.charAt(0);
+        return firstChar >= 'a' && firstChar <= 'z';
```
Comment (provenceee) [inline sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/utils/BaseHttpRouterUtils.java]: Maybe you can use Character.isLowerCase.

Comment (daizhenyu) [reply sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/utils/BaseHttpRouterUtils.java]: done

File: sermant-plugins/sermant-router/router-common/src/main/java/io/sermant/router/common/xds/XdsRouterHandler.java (hunk 6 above)
Comment (daizhenyu) [reply sermant-plugins/sermant-router/router-common/src/main/java/io/sermant/router/common/xds/XdsRouterHandler.java:204]: has checked for null

File: sermant-plugins/sermant-router/router-common/src/main/java/io/sermant/router/common/xds/XdsRouterHandler.java (hunk 5 above)
Comment (daizhenyu) [reply sermant-plugins/sermant-router/router-common/src/main/java/io/sermant/router/common/xds/XdsRouterHandler.java:198]: has checked for null

File: sermant-plugins/sermant-router/router-common/src/main/java/io/sermant/router/common/xds/XdsRouterHandler.java (hunk 4 above)
Comment (daizhenyu) [reply sermant-plugins/sermant-router/router-common/src/main/java/io/sermant/router/common/xds/XdsRouterHandler.java:192]: has checked for null

**Review by provenceee — APPROVED** [review]

**Review by lilai23 — APPROVED** [review]

**Review by hanbingleixue — APPROVED** [review]

**Pull request merged**